### PR TITLE
Branded images on Facebook shares

### PR DESCRIPTION
**Pull request by katebee**
## What does this change?

This adds an optional AB test that modifies the Facebook share url on all Facebook buttons on Guardian articles. Users with the modified url will post a different image to Facebook; one with the Guardian logo overlaid (see screenshots).
## What is the value of this and can you measure success?

This currently has zero customer participation to allow testing.
Hypotheses to be tested are that the branded overlay will improve brand recognition, click-through, and revisits.
Will track Facebook referrals to the modified url, and seven day returns for these users.
## Does this affect other platforms - Amp, Apps, etc?

This affects organic shares to Facebook.
Potential to affect any other sites that obtain Open Graph metadata from Guardian pages (Telegram messenger, Whatsapp?)
## Screenshots

**BEFORE**
before: normal facebook share

**AFTER**
after: facebook share with branded overlay
## Request for comment

@rich-nguyen @sndrs @gklopper

### CODE REVIEW

**Comment by paperboyo** [comment]
+1 (plus another one for a slight shadow that makes it stand out on white)

**Comment by katebee** [comment]
@desbo @gtrufitt - would you mind taking a look at this? Cheers :)

File: common/app/facebookimages/openGraphOverlay.scala [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package facebookimages
+
+import model.Article
+import views.support.{FacebookOpenGraphImage, ImgSrc}
+
+
+object openGraphOverlay {
+  def overlayImage(article: Article): Article = {
```
Comment (TBonnin) [inline common/app/facebookimages/openGraphOverlay.scala]: Should this method name hints it is returning an article. As is it looks more like it is returning an image.

File: common/app/facebookimages/openGraphOverlay.scala [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package facebookimages
+
+import model.Article
+import views.support.{FacebookOpenGraphImage, ImgSrc}
+
+
+object openGraphOverlay {
```
Comment (TBonnin) [inline common/app/facebookimages/openGraphOverlay.scala]: Why not making this object something like `ArticleWithOpenGraphOverlay` with an `apply` method that takes an article as argument?

**Comment by sndrs** [comment]
love it! is this a good place to use the strapline? 
screen shot 2016-05-18 at 10 18 31
not a blocker, just a query. 

also agree with @paperboyo too – an keyline/shadow would help visibility on light backgrounds

**Comment by paperboyo** [comment]
There is already a slight shadow there, @sndrs (may be a tad too wide as for my liking).

As to the design: it can be as complicated as we want, including pulling and compositioning cutouts, names, kickers, tones etc. per specific tone, type, contributor.

**Comment by katebee** [comment]
It is very easy to swap the image used for the overlay:

different overlay
(Obviously needs a higher res image but just a demo for @sndrs)

If anyone wants to try out the sandbox:
http://sandbox.imgix.com/view?url=https%3A%2F%2Fmedia-test.imgix.net%2F41cedfcc615f7c07207d3b9912ed8b7f113dbae4%2F0_0_2384_1431%2Fmaster%2F2384.jpg%3Fw%3D1200%26blend64%3DaHR0cHM6Ly9jbG91ZC5naXRodWJ1c2VyY29udGVudC5jb20vYXNzZXRzLzg2NzIzMy8xNTM1MzY0Ny8wMmM0YzllYy0xY2UyLTExZTYtOTJhZi02MDljYzZmMGNiZjEucG5n%26bm%3Dnormal%26ba%3Dbottom%26bw%3D1060%26bp%3D0%26h%3D632

The other logo is here: http://s14.postimg.org/4bp8p2xr5/White_logo_with_shadow.png

File: static/src/javascripts/projects/common/modules/experiments/tests/facebook-share-params.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
+define([
+    'common/utils/config'
+], function (
+    config
+) {
+    return function () {
+        this.id = 'FacebookShareParams';
+        this.start = '2016-06-06';
```
Comment (gtrufitt) [inline static/src/javascripts/projects/common/modules/experiments/tests/facebook-share-params.js]: Is this the right start date? You'll not be able to opt into the 0% test as this date is in June :)

**Comment by paperboyo** [comment]
These were my early tests, just to see what's possible (forget the actual design!):

branding test

**Comment by sndrs** [comment]
let's not get too deep into a discussion about what could happen on a PR that's doing one thing already :) (my comment may not have helped)

File: static/src/javascripts/projects/common/modules/experiments/tests/facebook-share-params.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        this.audienceCriteria = 'All users';
+        this.dataLinkNames = 'Open Graph';
+        this.idealOutcome = 'Increase in guardian traffic from facebook cards with image overlays';
+
+        this.canRun = function () {
+            return config.page.contentType == 'Article';
+        };
+
+        function alterCanonicalUrlSent(hrefContent, additionalParam) {
+            var urlhalves = hrefContent.split('&redirect_uri=');
```
Comment (gtrufitt) [inline static/src/javascripts/projects/common/modules/experiments/tests/facebook-share-params.js]: Just following this through but you're splitting at `&redirect_uri=` so you can get the `href=` value right? Do you need to urlEncode the `?` passed in here?

**Comment by gtrufitt** [comment]
👍 on the JS side bar those couple of comments, you might want to get someone else to review the scala :)

File: static/src/javascripts/projects/common/modules/experiments/tests/facebook-share-params.js (hunk 4 above)
Comment (katebee) [inline static/src/javascripts/projects/common/modules/experiments/tests/facebook-share-params.js]: No url encoding required 😄

File: static/src/javascripts/projects/common/modules/experiments/tests/facebook-share-params.js (hunk 3 above)
Comment (katebee) [inline static/src/javascripts/projects/common/modules/experiments/tests/facebook-share-params.js]: Cheers, have change it to today

File: common/app/facebookimages/openGraphOverlay.scala (hunk 2 above)
Comment (katebee) [inline common/app/facebookimages/openGraphOverlay.scala]: Thanks, have made this change! Any other thoughts on the Scala?

File: common/app/facebookimages/openGraphOverlay.scala [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  def overlayImage(article: Article): Article = {
+    article.copy(
+      content = article.content.copy(
+        metadata = article.content.metadata.copy(
+          opengraphPropertiesOverrides = {
+            val openGraph = article.content.metadata.opengraphProperties
+            openGraph ++ Map("og:image" -> ImgSrc(article.content.rawOpenGraphImage, FacebookOpenGraphImage, true))
```
Comment (davidfurey) [inline common/app/facebookimages/openGraphOverlay.scala]: There's no need to create a map with a single key:

```openGraph.updated("og:image", ImgSrc(article.content.rawOpenGraphImage, FacebookOpenGraphImage, true)))```
or
```openGraph + ("og:image" -> ImgSrc(article.content.rawOpenGraphImage, FacebookOpenGraphImage, true)))```

**Pull request merged**